src/env/config_paths.rs

```
@@ -1,154 +1,237 @@
+}
+
 const SYSCONF_DIR: &str = env!("SYSCONFDIR");
-const LOCALE_DIR: &str = env!("LOCALEDIR");
-const BIN_DIR: &str = env!("BINDIR");
-
-pub static CONFIG_PATHS: Lazy<ConfigPaths> = Lazy::new(|| {
-    // Read the current executable and follow all symlinks to it.
-    // OpenBSD has issues with `std::env::current_exe`, see gh-9086 and
-    // https://github.com/rust-lang/rust/issues/60560
-    let argv0 = PathBuf::from(std::env::args().next().unwrap());
-    let argv0 = if argv0.exists() {
-        argv0
-    } else {
-        std::env::current_exe().unwrap_or(argv0)
-    };
-    let argv0 = argv0.canonicalize().unwrap_or(argv0);
-    let mut paths = ConfigPaths::default();
-    let mut done = false;
-    let exec_path = get_executable_path(&argv0);
-    if let Ok(exec_path) = exec_path.canonicalize() {
-        FLOG!(
-            config,
-            format!("exec_path: {:?}, argv[0]: {:?}", exec_path, &argv0)
-        );
-        // TODO: we should determine program_name from argv0 somewhere in this file
+#[cfg(not(feature = "embed-data"))]
+const DOC_DIR: &str = env!("DOCDIR");
 
-        // Detect if we're running right out of the CMAKE build directory
-        if exec_path.starts_with(env!("FISH_BUILD_DIR")) {
-            let manifest_dir = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
-            FLOG!(
-                config,
-                "Running out of target directory, using paths relative to CARGO_MANIFEST_DIR:\n",
-                manifest_dir.display()
-            );
-            done = true;
-            paths = ConfigPaths {
-                data: Some(manifest_dir.join("share")),
-                sysconf: manifest_dir.join("etc"),
-                doc: manifest_dir.join("user_doc/html"),
-                bin: Some(exec_path.parent().unwrap().to_owned()),
-                locale: Some(manifest_dir.join("share/locale")),
-            }
+#[cfg(not(feature = "embed-data"))]
+enum ConfigPathKind {
+    BuildDirectory(/*manifest_dir=*/ PathBuf),
```

I get why you chose this naming and it does make sense, but it has the problem of inviting confusion since the "build" directory could easily be interpreted as `build/` or `target/`, which is probably why you also added the comment. `ManifestDirectory` might be a better choice. `RepoRootDirectory` could also work, as used in `ConfigPaths::from_source_tree`. Not sure about the best choice here, but I'd really like to have the info in the enum variant name instead of a comment which is much easier to miss. In any case, it would make sense to decide on what we want to call the directory and use it consistently throughout the codebase.

Another thing I noticed thinking about this is that we decide which paths to consider for data files based on fish's exec path, choosing the files in the repo if the exec path points into the repo. This works fine if using Cargo's default `target`, or a `build` directory inside the repo with CMake, but breaks if the build artifacts are located outside of the repo, either by using the `--target-dir` flag with Cargo or by letting CMake use a build dir that's not a subdir of the repo. I don't know if we want to change the behavior there. Fish could still try to access files from the repo in such scenarios, but it would have to check if the repo is still available, and it would be hard to decide whether to prefer files from the repo or in the installed location if both are available. Probably it's not worth the trouble to worry about such builds.

---

> Using "repo" also bothered me a bit

Let's use "Workspace" or "WorkspaceRoot" (or "WorkspacePath").
As used by Cargo, VSCode and Jujutsu.
The Cargo workspace root is probably always going to be the same as $(git rev-parse --show-toplevel).

> We might want to use code from `build-helper`

sounds good, nice.

> Ah yes, for some reason I thought the check is about the repo root

we should probably call it FISH_TARGET_DIR to move closer to Rust 

> I just tested Cargo's behavior

I raised https://github.com/rust-lang/cargo/issues/15962, maybe there
is enough want for this sort of thing

> but it would result in executables having access to data files
> in situations where they currently don't, e.g. when installed via
> cargo install.

but that'd only be relevant if we don't have embedding.
We can probably declare "cargo install --no-default-features" as unsupported
(I guess we should make it fail early rather than using random other runtime files...
except on OpenBSD where std::env::current_exe() won't work, so the compile-time paths are all we have).

---

> Let's use "Workspace" or "WorkspaceRoot" (or "WorkspacePath").

Good idea, that's easy to understand, hard to be confused about, and works in all scenarios.